File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
input = sys.stdin.readline

n, k = map(int, input().split())

items = [(0, 0)]
for _ in range(n):
    w, v = map(int, input().split())
    items.append((w, v)) # 무게, 가치

dp = [[0 for _ in range(k + 1)] for _ in range(n + 1)]

for i in range(1, n + 1):
    w, v = items[i]
    for j in range(1, k + 1):
        if j >= w:
            dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w] + v)
        else:
            dp[i][j] = dp[i - 1][j]

print(dp[n][k])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];

rl.on('line', input => {
  lines.push(input.trim())

  if (lines.length === parseInt(lines[0].split(' ')[0]) + 1) {
    rl.close()
  }
})

rl.on('close', () => {
  const [N, K] = lines.shift().split(' ').map(Number)
  let arr = []
  for (let i = 0; i < N; i++) {
    arr.push(lines[i].split(' ').map(Number))
  }
  
  let dp = Array.from({length: N + 1}, () => new Array(K + 1).fill(0))
  
  for (let i = 1; i < N + 1; i++) {
    const [W, V] = arr[i - 1]
    for (let j = 1; j < K + 1; j++) {
      if (j - W >= 0) {
        dp[i][j] = Math.max(dp[i - 1][j - W] + V, dp[i - 1][j]) 
      } else {
        dp[i][j] = dp[i - 1][j]
      }
    }
  }
  console.log(dp[N][K])
})